File: backend/src/handlers/put-job.mjs
// Create clients and set shared const values outside of the handler.

// Create a DocumentClient that represents the query to add an item
import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, PutCommand } from '@aws-sdk/lib-dynamodb';
import { Temporal } from '@js-temporal/polyfill';

// Get current time in a specific timezone (Paris for your Île-de-France jobs)
const now = Temporal.Now.zonedDateTimeISO('Europe/Paris');
const client = new DynamoDBClient({});
const ddbDocClient = DynamoDBDocumentClient.from(client);

// Get the DynamoDB table name from environment variables
const tableName = process.env.DYNAMODB_TABLE;

/**
 * A simple example includes a HTTP post method to add one item to a DynamoDB table.
 */
export const handler = async (event) => {
    if (event.httpMethod !== 'PUT') {
        throw new Error(`putJob only accepts PUT method, you tried: ${event.httpMethod} method.`);
    }
    // All log statements are written to CloudWatch
    console.info('received:', event);

    // Get id and name from the body of the request
    const body = JSON.parse(event.body);

    const job_title = body.job_title;
    const company_name = body.company_name;
    const link = body.link;
    const source = body.source;
    const year_of_experience = body.year_of_experience;
    const published_date = body.published_date;
    const description = body.description;
    const salary_range = body.salary_range;

    // Creates a new item, or replaces an old item with a new item
    // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/DynamoDB/DocumentClient.html#put-property
    var params = {
        TableName : tableName,
        Item: { job_title: job_title, company_name: company_name,
            link: link, source: source, year_of_experience: year_of_experience,
            published_date: published_date, description: description,
            salary_range: salary_range, created_at: now, updated_date: now }
    };

    try {
        const data = await ddbDocClient.send(new PutCommand(params));
        console.log("Success - item added or updated", data);
      } catch (err) {
        console.log("Error", err.stack);
      }

    const response = {
        statusCode: 200,
        body: JSON.stringify(body)
    };

    // All log statements are written to CloudWatch
    console.info(`response from: ${event.path} statusCode: ${response.statusCode} body: ${response.body}`);
    return response;
};

File: backend/src/api/put_job.py
"""
Lambda handler for creating/updating a job in DynamoDB table.
"""
import json
import logging
import os
import uuid
from datetime import datetime, timezone
from decimal import Decimal
from typing import Any, Dict, Optional

import boto3
from botocore.exceptions import ClientError
from pydantic import ValidationError

from src.shared.models import JobModel

# Configure logging
logger = logging.getLogger()
logger.setLevel(logging.INFO)

# Create DynamoDB client outside handler for connection reuse
# Check if we're running locally with DynamoDB Local
dynamodb_endpoint = os.environ.get('DYNAMODB_ENDPOINT')
if dynamodb_endpoint:
    # Use local DynamoDB endpoint
    dynamodb = boto3.resource('dynamodb', endpoint_url=dynamodb_endpoint)
else:
    # Use AWS DynamoDB
    dynamodb = boto3.resource('dynamodb')


def handler(event: Dict[str, Any], context: Any) -> Dict[str, Any]:
    """
    Lambda handler to create/update a job in DynamoDB table.
    
    Args:
        event: API Gateway event
        context: Lambda context
        
    Returns:
        API Gateway response
    """
    if event.get('httpMethod') != 'PUT':
        raise ValueError(f"putJob only accepts PUT method, you tried: {event.get('httpMethod')}")
    
    logger.info(f"Received event: {json.dumps(event)}")
    
    # Get table name from environment
    table_name = os.environ.get('DYNAMODB_TABLE')
    if not table_name:
        raise ValueError("DYNAMODB_TABLE environment variable is required")
    
    table = dynamodb.Table(table_name)
    
    # Parse request body
    try:
        body = json.loads(event.get('body', '{}'))
    except json.JSONDecodeError:
        return {
            'statusCode': 400,
            'body': json.dumps({'error': 'Invalid JSON in request body'})
        }
    
    # Validate request body using Pydantic
    try:
        job_data = JobModel(**body)
    except ValidationError as e:
        logger.error(f"Validation error: {e}")
        return {
            'statusCode': 400,
            'body': json.dumps({'error': 'Invalid request data', 'details': e.errors()})
        }
    
    # Generate UUID for the job and current timestamp
    job_id = str(uuid.uuid4())
    now = datetime.now(timezone.utc).isoformat().replace('+00:00', 'Z')
    
    # Prepare item for DynamoDB
    item = {
        'id': job_id,
        'job_title': job_data.job_title,
        'company_name': job_data.company_name,
        'link': job_data.link,
        'source': job_data.source,
        'year_of_experience': job_data.year_of_experience,
        'published_date': job_data.published_date,
        'created_at': now,
        'updated_at': now
    }
    
    # Add optional fields if provided
    if job_data.description:
        item['description'] = job_data.description
    if job_data.salary_range:
        item['salary_range'] = Decimal(str(job_data.salary_range))
    
    try:
        # Put item in DynamoDB table
        table.put_item(Item=item)
        logger.info(f"Successfully created/updated job with id: {job_id}")
        
    except ClientError as e:
        logger.error(f"Error putting item: {e}")
        return {
            'statusCode': 500,
            'body': json.dumps({'error': 'Failed to create/update job'})
        }
    
    response_body = {
        'statusCode': 200,
        'body': json.dumps({**body, 'id': job_id, 'created_at': now, 'updated_at': now})
    }
    
    logger.info(f"Response from {event.get('path')}: statusCode: {response_body['statusCode']}")
    return response_body